test/utility/file.c

```diff
@@ -0,0 +1,534 @@
+{
+    struct copy_data *d;
+    char dest_path[PATH_MAX];
+
+    WT_UNUSED(path);
+    d = (struct copy_data *)user_data;
+
+    /* No need to do anything for the top-level directory. This is handled elsewhere. */
+    if (info->rel_path == NULL || strcmp(info->rel_path, ".") == 0)
+        return;
+
+    /* Check if we need to switch to using links. */
+    if (d->opts->link && d->link_depth < 0)
+        if (strncmp(d->opts->link_if_prefix, info->base_name, strlen(d->opts->link_if_prefix)) == 0)
+            d->link_depth = info->depth;
+
+    /* Otherwise we create a new directory. */
+    testutil_snprintf(
+      dest_path, sizeof(dest_path), "%s" DIR_DELIM_STR "%s", d->dest, info->rel_path);
+    testutil_assert_errno(mkdir(dest_path, info->stat.st_mode) == 0);
+}
+
+/*
+ * copy_on_directory_leave --
+ *     Worker for copying a directory.
+ */
+static void
+copy_on_directory_leave(const char *path, const file_info_t *info, void *user_data)
+{
+    struct copy_data *d;
+#if defined(_WIN32)
+    struct _utimbuf t;
+#elif defined(__linux__)
+    struct timeval times[2];
+#else
+    struct utimbuf t;
+#endif
+    char dest_path[PATH_MAX];
+
+    WT_UNUSED(path);
+    d = (struct copy_data *)user_data;
+
+    if (info->depth <= d->link_depth)
+        d->link_depth = -1;
+
+    /* Preserve the timestamps. */
+    if (d->opts->preserve) {
+        testutil_snprintf(dest_path, sizeof(dest_path), "%s" DIR_DELIM_STR "%s", d->dest,
+          info->rel_path == NULL ? "" : info->rel_path);
+#if defined(_WIN32)
+        t.actime = info->stat.st_atime;
+        t.modtime = info->stat.st_mtime;
+        testutil_assert_errno(_utime(dest_path, &t) == 0);
+#elif defined(__linux__)
+        times[0].tv_sec = info->stat.st_atim.tv_sec;
+        times[0].tv_usec = info->stat.st_atim.tv_nsec / 1000;
+        times[1].tv_sec = info->stat.st_mtim.tv_sec;
+        times[1].tv_usec = info->stat.st_mtim.tv_nsec / 1000;
+        testutil_assert_errno(utimes(dest_path, times) == 0);
+#else
+        t.actime = info->stat.st_atime;
+        t.modtime = info->stat.st_mtime;
+        testutil_assert_errno(utime(dest_path, &t) == 0);
+#endif
+    }
+}
+
+/*
+ * testutil_copy --
+ *     Recursively copy a file or a directory tree. Fail the test on error.
+ */
+void
+testutil_copy(const char *source, const char *dest)
+{
+    testutil_copy_ext(source, dest, NULL);
+}
+
+/*
+ * testutil_copy_ext --
+ *     Recursively copy a file or a directory tree. Fail the test on error. With extra options.
+ */
+void
+testutil_copy_ext(const char *source, const char *dest, const WT_FILE_COPY_OPTS *opts)
+{
+    struct copy_data data;
+    struct stat source_stat, dest_stat;
+    WT_DECL_RET;
+    WT_FILE_COPY_OPTS default_opts;
+    const char *path;
+    bool dest_exists;
+    bool is_dest_dir;
+    bool is_source_dir;
+    glob_t g;
+    size_t i;
+
+    if (opts == NULL) {
+        memset(&default_opts, 0, sizeof(default_opts));
+        opts = &default_opts;
+    }
```

Comment:
Nit: could also make a `static const WT_FILE_COPY_OPTS default_opts;` to avoid the memset.